Mathlib/Tactic/CC/Addition.lean

```diff
@@ -0,0 +1,1892 @@
+
+See paper: Congruence Closure for Intensional Type Theory. -/
+partial def isCongruent (e₁ e₂ : Expr) : CCM Bool := do
+  let .app f a := e₁ | failure
+  let .app g b := e₂ | failure
+  if (← getEntry e₁).any Entry.fo then
+    e₁.withApp fun f₁ args₁ =>
+    e₂.withApp fun f₂ args₂ => do
+      if ha : args₁.size = args₂.size then
+        for hi : i in [:args₁.size] do
+          if (← getRoot (args₁[i]'hi.2)) != (← getRoot (args₂[i]'(ha.symm ▸ hi.2))) then
+            return false
+        if f₁ == f₂ then return true
+        else if (← getRoot f₁) != (← getRoot f₂) then
+          -- `f₁` and `f₂` are not equivalent
+          return false
+        else if ← pureIsDefEq (← inferType f₁) (← inferType f₂) then
+          return true
+        else return false
+      else return false
+  else
+    -- Given `e₁ := f a`, `e₂ := g b`
+    if (← getRoot a) != (← getRoot b) then
+      -- `a` and `b` are not equivalent
+      return false
+    else if (← getRoot f) != (← getRoot g) then
+      -- `f` and `g` are not equivalent
+      return false
+    else if ← pureIsDefEq (← inferType f) (← inferType g) then
+      /- Case 1: `f` and `g` have the same type, then we can create a congruence proof for
+         `HEq (f a) (g b)` -/
+      return true
+    else if f.isApp && g.isApp then
+      -- Case 2: `f` and `g` are congruent
+      isCongruent f g
+    else
+      /-
+      f and g are not congruent nor they have the same type.
+      We can't generate a congruence proof in this case because the following lemma
+        `hcongr : HEq f₁ f₂ → HEq a₁ a₂ → HEq (f₁ a₁) (f₂ a₂)`
+      is not provable.
+      Remark: it is also not provable in MLTT, Coq and Agda (even if we assume UIP).
+      -/
+      return false
+
+def mkCongruencesKey (e : Expr) : CCM CongruencesKey := do
+  let .app f a := e | failure
+  if (← getEntry e).any Entry.fo then
+    -- first-order case, where we do not consider all partial applications
+    e.withApp fun fn args => do
+      return .fo (← getRoot fn) (← args.mapM getRoot)
+  else
+    return .ho (← getRoot f) (← getRoot a)
+
+def mkSymmCongruencesKey (lhs rhs : Expr) : CCM SymmCongruencesKey := do
+  let lhs ← getRoot lhs
+  let rhs ← getRoot rhs
+  if hash lhs > hash rhs then return { h₁ := rhs, h₂ := lhs } else return { h₁ := lhs, h₂ := rhs }
+
+def mkCCCongrTheorem (e : Expr) : CCM (Option CCCongrTheorem) := do
+  let fn := e.getAppFn
+  let nargs := e.getAppNumArgs
+  let cache ← getCache
+
+  -- Check if `{ fn, nargs }` is in the cache
+  let key₁ : CCCongrTheoremKey := { fn, nargs }
+  if let some it₁ := cache.findEntry? key₁ then
+    return it₁.2
+
+  -- Try automatically generated congruence lemma with support for heterogeneous equality.
+  let lemm ← mkCCHCongrWithArity fn nargs
+
+  if let some lemm := lemm then
+    modifyCache fun ccc => ccc.insert key₁ (some lemm)
+    return lemm
+
+  -- cache failure
+  modifyCache fun ccc => ccc.insert key₁ none
+  return none
```

Comment:
This is literally the same as the body of `mkCCHCongrTheorem`. Could you switch the order and call `mkCCHCongrTheorem` here?